stock_quant_constraint/models/stock_quant.py

```diff
@@ -0,0 +1,206 @@
+        )
+        for quant in self:
+            reserved_quantity = quant.reserved_quantity or 0.0
+            location = quant.location_id
+            if reserved_quantity > 0.0 and location.usage != "internal":
+                raise ValidationError(
+                    _(
+                        "Invalid attempt to reserve product %s"
+                        " for a location with usage %s."
+                    )
+                    % (quant.product_id.display_name, location.usage,)
+                )
+            if location.usage != "internal":
+                continue
+            # Compare reserved quantity in quant with moves for the location.
+            parameters = {
+                "location_id": self.location_id.id,
+                "product_id": self.product_id.id,
+                "lot_id": self.lot_id.id or 0,
+                "package_id": self.package_id.id or 0,
+                "owner_id": self.owner_id.id or 0,
+            }
+            self.env.cr.execute(RESERVED_STATEMENT, parameters)
+            reserved_in_moves = self.env.cr.fetchone()[0]
+            _logger.info(
+                _("Reserved in quants %s, reserved in moves %s for keys %s"),
+                reserved_quantity,
+                reserved_in_moves,
+                str(parameters),
+            )
+            comparison = float_compare(
+                reserved_quantity, reserved_in_moves, precision_digits=precision
+            )
+            if comparison == 0:
+                continue
+            # We detected a mismatch between moves and quants.
+            raise ValidationError(
+                _(
+                    "Mismatch between reserved product quantity according to"
+                    " quants %s and accourding to moves %s for product %s"
+                    " in location %s."
+                )
+                % (
+                    reserved_quantity,
+                    reserved_in_moves,
+                    quant.product_id.display_name,
+                    location.complete_name,
+                )
+            )
+
+    @api.model
+    def cron_check_quant_consistency(self):
+        """Check consistency of quants when compared with stock move lines."""
+        cursor = self.env.cr
+        # First check wether quants have unique keys.
+        quants_unique = True
+        cursor.execute(UNIQUE_QUANTS_STATEMENT)
+        for row in cursor.fetchall():
+            quants_unique = False
+            _logger.error(_("Duplicate quant keys %s"), str(row))
+        if not quants_unique:
+            raise UserError(_("Cron to check quants aborted because of duplicat keys"))
+        # Continue checking quants for mismatches.
+        cursor.execute(STOCK_STATEMENT)
+        for row in cursor.fetchall():
+            _logger.error(_("Mismatch in stock %s"), str(row))
+            self._do_repair(row)
+
+    def _do_repair(self, row):
+        """Repair (or create) stock_quant, based on stock moves."""
```

Comment:
@NL66278 Well the docstring looks promising!

I have quite some experience in this problem. The biggest issue is of course failed to reserve/unreserve. I have never quite traced it but I feel it has to do with stock.inventory validation. In general the total qty is correct, but the movelines are not. The other one is within a failed transaction, again never quite tracked it, but the quant gets committed and the move rolled back, which of course breaks 2 quants.

I am very interested in this problem and have some database full of inconsistencies to test. V12 and V14 only mind you but looks pretty easy backport for me once you have something. Just ping when ready.

Reply:
@gdgellatly I did not have time to work further on this, but I have now asked one of our programmers to get involved.